src/workerd/jsg/url-test.c++

```diff
@@ -100,5 +102,1764 @@ KJ_TEST("Search params (2)") {
   KJ_ASSERT(kj::str(params) == "a=z");
 }
 
+// ======================================================================================
+
+KJ_TEST("URLPattern - processInit Default") {
+  KJ_SWITCH_ONEOF(UrlPattern::processInit(UrlPattern::Init {})) {
+    KJ_CASE_ONEOF(result, UrlPattern::Init) {
+      KJ_ASSERT(result.baseUrl == kj::none);
+      KJ_ASSERT(result.protocol == kj::none);
+      KJ_ASSERT(result.username == kj::none);
+      KJ_ASSERT(result.password == kj::none);
+      KJ_ASSERT(result.hostname == kj::none);
+      KJ_ASSERT(result.port == kj::none);
+      KJ_ASSERT(result.pathname == kj::none);
+      KJ_ASSERT(result.search == kj::none);
+      KJ_ASSERT(result.hash == kj::none);
+    }
+    KJ_CASE_ONEOF(msg, kj::String) {
+      KJ_FAIL_ASSERT("Default processInit failed", msg);
+    }
+  }
+}
+
+KJ_TEST("URLPattern - processInit PATTERN mode") {
+  KJ_SWITCH_ONEOF(UrlPattern::processInit(UrlPattern::Init {
+    // Since we're using PATTERN mode here (the default), the values
+    // for each field will not be canonicalized.
+    .protocol = kj::str("something"),
+    .username = kj::str("something"),
+    .password = kj::str("something"),
+    .hostname = kj::str("something"),
+    .port = kj::str("something"),
+    .pathname = kj::str("something"),
+    .search = kj::str("something"),
+    .hash = kj::str("something"),
+  })) {
+    KJ_CASE_ONEOF(result, UrlPattern::Init) {
+      KJ_ASSERT(result.baseUrl == kj::none);
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.protocol) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.hostname) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.pathname) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.username) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.password) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.port) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.search) == "something");
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.hash) == "something");
+    }
+    KJ_CASE_ONEOF(msg, kj::String) {
+      KJ_FAIL_ASSERT("processInit failed", msg);
+    }
+  }
+}
+
+KJ_TEST("URLPattern - processInit PATTERN mode") {
+  KJ_SWITCH_ONEOF(UrlPattern::processInit(UrlPattern::Init {
+    // Since we're using PATTERN mode here (the default), the values
+    // for each field will not be canonicalized.
+    .protocol = kj::str("something"),
+    .username = kj::str("something"),
+    .password = kj::str("something"),
+    .hostname = kj::str("something"),
+    .pathname = kj::str("something"),
+    .hash = kj::str("something"),
+  }, UrlPattern::ProcessInitOptions {
+    .port = "something"_kj,
+    .search = "something"_kj,
+  })) {
+    KJ_CASE_ONEOF(result, UrlPattern::Init) {
+      KJ_ASSERT(result.baseUrl == kj::none);
+      KJ_ASSERT(KJ_ASSERT_NONNULL(result.protocol) == "something");
```

Comment:
I think these tests would benefit from the same data driven approach you have in javascript tests. no sure how pleasant this would be in c++, but maybe it is doable?

Reply:
Doable, yes, I'll save that refactor to a separate PR tho.